compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,320 @@
+
+        let ssa_locals = SsaLocals::new(body);
+        if simplify_static_switches(tcx, body, &ssa_locals) {
+            simplify::remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip_all, ret)]
+fn simplify_static_switches<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &mut Body<'tcx>,
+    ssa_locals: &SsaLocals,
+) -> bool {
+    let dominators = body.basic_blocks.dominators();
+    let predecessors = body.basic_blocks.predecessors();
+    let mut discriminants = FxHashMap::default();
+    let mut static_switches = FxHashMap::default();
+    let mut borrowed_locals =
+        MaybeBorrowedLocals.into_engine(tcx, body).iterate_to_fixpoint().into_results_cursor(body);
+    for (switched, rvalue, location) in ssa_locals.assignments(body) {
+        let Rvalue::Discriminant(discr) = rvalue else {
+            continue
+        };
+
+        borrowed_locals.seek_after_primary_effect(location);
+        // If `discr` was borrowed before its discriminant was assigned to `switched`,
+        // or if it was borrowed in the assignment, we cannot optimize.
+        if borrowed_locals.contains(discr.local) {
+            debug!("The aggregate: {discr:?} was borrowed before its discriminant was read");
+            continue;
+        }
+
+        let Location { block, statement_index } = location;
+        let mut finder = MutatedLocalFinder { local: discr.local, mutated: false };
+        for (statement_index, statement) in body.basic_blocks[block]
+            .statements
+            .iter()
+            .enumerate()
+            .take_while(|&(index, _)| index != statement_index)
+        {
+            finder.visit_statement(statement, Location { block, statement_index });
+        }
+
+        if finder.mutated {
+            debug!("The aggregate: {discr:?} was mutated before its discriminant was read");
+            continue;
+        }
+
+        // If `switched` is borrowed by the time we actually switch on it, we also cannot optimize.
+        borrowed_locals.seek_to_block_end(block);
+        if borrowed_locals.contains(switched) {
+            debug!("The local: {switched:?} was borrowed before being switched on");
+            continue;
+        }
+
+        discriminants.insert(
+            switched,
+            Discriminant {
+                block,
+                discr: discr.local,
+                exclude: if ssa_locals.num_direct_uses(switched) == 1 {
+                    // If there is only one direct use of `switched` we do not need to keep
+                    // it around because the only use is in the switch.
+                    Some(statement_index)
+                } else {
+                    None
+                },
+            },
+        );
+    }
+
+    if discriminants.is_empty() {
+        debug!("No SSA locals were assigned a discriminant");
+        return false;
+    }
+
+    for (switched, Discriminant { discr, block, exclude }) in discriminants {
```

Comment:
`discriminants` is a hashmap. Iteration order is unstable. Should `discriminants` be a Vec or pairs?